planner/cascades/transformation_rules.go

```diff
@@ -132,6 +133,87 @@ func (r *PushSelDownTableScan) OnTransform(old *memo.ExprIter) (newExprs []*memo
 	return []*memo.GroupExpr{selExpr}, true, false, nil
 }
 
+// PushSelDownIndexScan pushes a Selection down to IndexScan.
+type PushSelDownIndexScan struct {
+	baseRule
+}
+
+// NewRulePushSelDownIndexScan creates a new Transformation PushSelDownIndexScan.
+// The pattern of this rule is `Selection -> IndexScan`.
+func NewRulePushSelDownIndexScan() Transformation {
+	rule := &PushSelDownIndexScan{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandSelection,
+		memo.EngineTiKVOnly,
+		memo.NewPattern(memo.OperandIndexScan, memo.EngineTiKVOnly),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `Selection -> IndexScan` to:
+//   `IndexScan(with a new access range)` or
+//   `Selection -> IndexScan(with a new access range)`
+//	 or just keep the two GroupExprs unchanged.
+func (r *PushSelDownIndexScan) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	sel := old.GetExpr().ExprNode.(*plannercore.LogicalSelection)
+	is := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalIndexScan)
+	if len(is.IdxCols) == 0 {
+		return nil, false, false, nil
+	}
+	conditions := sel.Conditions
+	if is.AccessConds != nil {
+		// If we have already pushed some conditions down here,
+		// we merge old AccessConds with new conditions,
+		// to make sure this rule can be applied more than once.
+		conditions = make([]expression.Expression, len(sel.Conditions)+len(is.AccessConds))
+		copy(conditions, sel.Conditions)
+		copy(conditions[len(sel.Conditions):], is.AccessConds)
+	}
+	res, err := ranger.DetachCondAndBuildRangeForIndex(is.SCtx(), conditions, is.IdxCols, is.IdxColLens)
+	if err != nil {
+		return nil, false, false, err
+	}
+	if len(res.AccessConds) == len(is.AccessConds) {
```

Comment:
They are not the same. This only happens if we have more than one selection. For example:
```
select a, b from (select a, b from t where b > 1) as t1 where a > 1
```
We have an index `b`. Because we are using a bottom-up searching method, we will push down `b > 1` into the `IndexScan` at the first time. Then we will apply this rule again with the condition `a > 1`, and we find that this condition cannot be push down to this index. So we come to this branch with `len(res.AccessConds) == len(is.AccessConds)`.

Reply:
You are right.